README.txt
pleiades.openlayers Package Readme
=========================

Overview
--------

OpenLayers support for Pleiades


docs/HISTORY.txt
Changelog for pleiades.openlayers

    (name of developer listed in brackets)

pleiades.openlayers - 0.1 Unreleased

    - Initial package structure.
      [zopeskel]



pleiades/openlayers/tests.py
import unittest

from zope.testing import doctestunit
from zope.component import testing
from Testing import ZopeTestCase as ztc

from Products.Five import zcml
from Products.Five import fiveconfigure
from Products.PloneTestCase import PloneTestCase as ptc
from Products.PloneTestCase.layer import PloneSite
ptc.setupPloneSite()

import pleiades.openlayers

class TestCase(ptc.PloneTestCase):
    class layer(PloneSite):
        @classmethod
        def setUp(cls):
            fiveconfigure.debug_mode = True
            zcml.load_config('configure.zcml',
                             pleiades.openlayers)
            fiveconfigure.debug_mode = False

        @classmethod
        def tearDown(cls):
            pass


def test_suite():
    return unittest.TestSuite([

        # Unit tests
        #doctestunit.DocFileSuite(
        #    'README.txt', package='pleiades.openlayers',
        #    setUp=testing.setUp, tearDown=testing.tearDown),

        #doctestunit.DocTestSuite(
        #    module='pleiades.openlayers.mymodule',
        #    setUp=testing.setUp, tearDown=testing.tearDown),


        # Integration tests that use PloneTestCase
        #ztc.ZopeDocFileSuite(
        #    'README.txt', package='pleiades.openlayers',
        #    test_class=TestCase),

        #ztc.FunctionalDocFileSuite(
        #    'browser.txt', package='pleiades.openlayers',
        #    test_class=TestCase),

        ])

if __name__ == '__main__':
    unittest.main(defaultTest='test_suite')


pleiades/openlayers/proj.py
import geojson
from pyproj import Proj


PROJ_900913 = """
+proj=merc +a=6378137 +b=6378137
+lat_ts=0.0 +lon_0=0.0 +x_0=0.0 +y_0=0 +k=1.0
+units=m +nadgrids=@null +no_defs
"""


class Transform(object):
    
    def __init__(self, proj_defn):
        self.proj = Proj(proj_defn)

    def __call__(self, ob, inverse=False, object_hook=None):
        """Transform an object which provides the geo interface from 
        geographic to projected coordinates.
        
        If the inverse keyword is True, pyproj computes the inverse transform.

        If an object_hook callable is provided (as in geojson), instances of
        a particular class will be returned. Otherwise instances of geojson
        geometry classes are returned.
        """
        geo = getattr(ob, '__geo_interface__', ob)
        geo = geo.get('geometry', geo)
        constructor = object_hook or geojson.GeoJSON.to_instance
        return constructor(self.transform_geom(geo, inverse))

    def transform_coords1(self, coords, inverse=False):
        x, y = self.proj(coords[0], coords[1], **dict(inverse=inverse))
        return (x, y)

    def transform_coords2(self, coords, inverse=False):
        values = zip(*coords)
        xs, ys = self.proj(values[0], values[1], **dict(inverse=inverse))
        return tuple(zip(xs, ys))

    def transform_coords3(self, coords, inverse=False):
        return tuple([self.transform_coords2(r, inverse) for r in coords])
        
    def transform_coords4(self, coords, inverse=False):
        return tuple([self.transform_coords3(p, inverse) for p in coords])
    
    def transform_geom(self, geo, inverse=False):
        gtype = geo['type']
        coords = geo['coordinates']
        if gtype == 'Point':
            result = self.transform_coords1(coords, inverse)
        elif gtype in ['LineString', 'LinearRing', 'MultiPoint']:
            result = self.transform_coords2(coords, inverse)
        elif gtype in ['Polygon', 'MultiLineString']:
            result = self.transform_coords3(coords, inverse)
        elif gtype == 'MultiPolygon':
            result = self.transform_coords4(coords, inverse)
        else:
            raise NotImplemented, "No geometry collections in this version"
        return dict(type=gtype, coordinates=result)


pleiades/openlayers/browser.py
import os
import geojson
from shapely.geometry import asShape
from zope.publisher.browser import BrowserPage
from zope.event import notify
from zgeo.geographer.geo import ObjectGeoreferencedEvent
from zgeo.geographer.interfaces import IGeoreferenced
from pleiades.openlayers.proj import Transform, PROJ_900913


class OLSphericalMercatorJS(BrowserPage):
    """Returns coordinates projected to Spherical Mercator for use with 
    Google Maps.
    """

    def __init__(self, context, request):
        self.context = context
        self.request = request
        self.transform = Transform(PROJ_900913)

    def __call__(self):
        response = self.request.response
        geo = IGeoreferenced(self.context)
        where = self.transform(geo)
        if geo.type == 'Point':
            centroid = where
        else:
            shape = asShape(geo)
            centroid = self.transform(shape.centroid)
        response.setHeader('Content-Type', 'text/javascript')
        return """
// Javascript summary of a Pleiades entity
var pleiades_oljs = %s;
        """ % geojson.dumps(
                dict(
                    uid=self.context.UID(),
                    where=where,
                    centroid=centroid
                    )
                )

GMAPS_KEY = os.environ.get('GMAPS_KEY', '')

class GoogleAPIKey(BrowserPage):

    def __call__(self):
        return GMAPS_KEY


class EditGeo(BrowserPage):

    def __init__(self, context, request):
        self.context = context
        self.request = request
        self.transform = Transform(PROJ_900913)

    def __call__(self):
        """Request parameters are 'type' and 'coordinates'."""
        request = self.request
        response = request.response
        try:
            gtype = request.form.get('type')
            coords = request.form.get('coordinates')
            data = '{"type": "%s", "coordinates": %s}' % (gtype, coords)
            obj = geojson.loads(data, object_hook=geojson.GeoJSON.to_instance)
        except:
            raise
            response.setStatus(400)
            return "Input geometry is not acceptable"
        
        # Input is 900913, transform back to lon/lat
        result = self.transform(obj, inverse=True)        
        g = IGeoreferenced(self.context)
        g.setGeoInterface(result.type, result.coordinates)
        notify(ObjectGeoreferencedEvent(g))

        if request.get('HTTP_REFERER'):
            response.redirect(request.get('HTTP_REFERER').split('?')[0] + '?portal_status_message=Changes%20saved.')
        else:
            response.setStatus(200)
            return "Geometry edited successfully"



pleiades/openlayers/browser/views.py
import os
import geojson
from shapely.geometry import asShape
from zope.publisher.browser import BrowserPage
from zope.event import notify
from zgeo.geographer.geo import ObjectGeoreferencedEvent
from zgeo.geographer.interfaces import IGeoreferenced
from pleiades.openlayers.proj import Transform, PROJ_900913


class OLSphericalMercatorJS(BrowserPage):
    """Returns coordinates projected to Spherical Mercator for use with 
    Google Maps.
    """

    def __init__(self, context, request):
        self.context = context
        self.request = request
        self.transform = Transform(PROJ_900913)

    def __call__(self):
        response = self.request.response
        geo = IGeoreferenced(self.context)
        where = self.transform(geo)
        if geo.type == 'Point':
            centroid = where
        else:
            shape = asShape(geo)
            centroid = self.transform(shape.centroid)
        context_url = self.context.absolute_url()
        response.setHeader('Content-Type', 'text/javascript')
        return """
// Javascript summary of a Pleiades entity
var url_marker_gold = "%s/++resource++marker_gold.png";
var url_marker_shadow = "%s/++resource++marker_shadow.png";
var pleiades_oljs = %s;
        """ % (
                context_url,
                context_url,
                geojson.dumps(
                dict(
                    uid=self.context.UID(),
                    where=where,
                    centroid=centroid
                    )
                )
              )

GMAPS_KEY = os.environ.get('GMAPS_KEY', '')

class GoogleAPIKey(BrowserPage):

    def __call__(self):
        return GMAPS_KEY


class EditGeo(BrowserPage):

    def __init__(self, context, request):
        self.context = context
        self.request = request
        self.transform = Transform(PROJ_900913)

    def __call__(self):
        """Request parameters are 'type' and 'coordinates'."""
        request = self.request
        response = request.response
        try:
            gtype = request.form.get('type')
            coords = request.form.get('coordinates')
            data = '{"type": "%s", "coordinates": %s}' % (gtype, coords)
            obj = geojson.loads(data, object_hook=geojson.GeoJSON.to_instance)
        except:
            raise
            response.setStatus(400)
            return "Input geometry is not acceptable"
        
        # Input is 900913, transform back to lon/lat
        result = self.transform(obj, inverse=True)        
        g = IGeoreferenced(self.context)
        g.setGeoInterface(result.type, result.coordinates)
        notify(ObjectGeoreferencedEvent(g))

        if request.get('HTTP_REFERER'):
            response.redirect(request.get('HTTP_REFERER').split('?')[0] + '?portal_status_message=Changes%20saved.')
        else:
            response.setStatus(200)
            return "Geometry edited successfully"

